rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,228 @@
+
+```typescript
+server.plugins.encrypted_saved_objects.registerType({
+  type: 'server-action',
+  attributesToEncrypt: new Set(['credentials', 'apiKey']),
+  // Optional list of attributes that might not contain user specified data.
+  attributesToExcludeFromAAD: new Set(['temporaryState']),
+});
+```
+
+Use the same API to create saved objects with encrypted attributes as for any other saved object type:
+
+```typescript
+const savedObject = await server.savedObjects
+  .getScopedSavedObjectsClient(request)
+  .create('server-action', { 
+    name: 'my-server-action',
+    credentials: { username: 'some-user', password: 'some-password' },
+    apiKey: 'dGhpcyBpcyBub3QgYSByZWFsIHRva2VuIGJ1dCBpdCBpcyBvb'
+  });
+
+// savedObject = { 
+//   id: 'dd9750b9-ef0a-444c-8405-4dfcc2e9d670',
+//   type: 'server-action',
+//   name: 'my-server-action',
+// };
+
+```
+
+Use dedicated method to retrieve saved object with decrypted attributes on behalf of Kibana internal user:
+
+```typescript
+const savedObject = await server.plugins.encrypted_saved_objects.getDecryptedAsInternalUser(
+  'server-action',
+  'dd9750b9-ef0a-444c-8405-4dfcc2e9d670'
+);
+
+// savedObject = { 
+//   id: 'dd9750b9-ef0a-444c-8405-4dfcc2e9d670',
+//   type: 'server-action',
+//   name: 'my-server-action',
+//   credentials: { username: 'some-user', password: 'some-password' },
+//   apiKey: 'dGhpcyBpcyBub3QgYSByZWFsIHRva2VuIGJ1dCBpdCBpcyBvb',
+// };
+```
+
+# Motivation
+
+Main motivation is the storage and usage of third-party credentials for use with
+the action service to do notifications. Also perform other types integrations,
+call webhooks using tokens.
+
+# Detailed design
+
+In order for this to be in basic it needs to be done as a wrapper around the
+saved object client. This can be added from the `x-pack` plugin.
+
+## General
+
+To be able to manage saved objects with encrypted attributes from any plugin one should
+do the following:
+
+1. Define `encrypted_saved_objects` plugin as a dependency.
+2. Add attributes to be encrypted in `mappings.json` file for the respective saved object type. These attributes should
+always have a `binary` type since they'll contain encrypted content as a `Base64` encoded string and should never be
+searchable or analyzed. This makes defining of attributes that require encryption explicit and auditable, and significantly
+simplifies implementation:
+```json
+{
+ "server-action": {
+   "properties": {
+     "name": { "type": "keyword" },
+     "credentials": { "type": "binary" },
+     "apiKey": { "type": "binary" }
+   }
+ }
+}
+```
+3. Register saved object type and attributes that should be encrypted with `encrypted_saved_objects` plugin:
+```typescript
+server.plugins.encrypted_saved_objects.registerType({
+  type: 'server-action',
+  attributesToEncrypt: new Set(['credentials', 'apiKey']),
+  // Optional list of attributes that might not contain user specified data.
+  attributesToExcludeFromAAD: new Set(['temporaryState']),
```

Comment:
This should be removed from this example, right? As far as I can tell, this property is irrelevant here.

Reply:
I'll keep it here, just change example mapping to cover this case.